### PR TITLE
Allow casting one-element Tensors onto Python scalars

### PR DESCRIPTION
**Pull request by jytug**
Regarding #1129

### CODE REVIEW

File: torch/tensor.py [hunk 1]
```diff
@@ -371,6 +371,24 @@ def __invert__(self):
     def __hash__(self):
         return id(self)
 
+    def __int__(self):
+        if self.numel() == 1:
+            return int(self.storage()[0])
+        raise TypeError("only 1-element tensors can be converted "
+                        "to Python scalars")
```
Comment (gchanan) [inline torch/tensor.py]: should we be explicit about the type?  "Python scalars" aren't actually a defined type, right?

**Review by gchanan — COMMENTED** [review]
Does this work for variables as well?  (You may have to change variable.py).

Also, some tests?

File: torch/tensor.py (hunk 1 above)
Comment (jytug) [reply torch/tensor.py]: I was inspired by `numpy`'s error message:
```
TypeError: only length-1 arrays can be converted to Python scalars
```
Of course it won't harm for `torch` to be more verbose

File: torch/tensor.py [hunk 2]
```diff
@@ -371,6 +371,24 @@ def __invert__(self):
     def __hash__(self):
         return id(self)
 
+    def __int__(self):
+        if self.numel() == 1:
+            return int(self.storage()[0])
```
Comment (apaszke) [inline torch/tensor.py]: This is incorrect if tensor has non-zero data offset. I think the safest way would be to do `self[(0,) * self.ndim()]`

Comment (jytug) [reply torch/tensor.py]: Flew over my head:

```
>> x = torch.Tensor([1,2])
>> y = x[1:]
>> int(y)
1
```

**Review by soumith — APPROVED** [review]

**Comment by soumith** [comment]
thank you Filip!

**Comment by EntilZha** [comment]
I can make a separate issue if it would be ebtter, but this might be a good place to start since this change seems like it might have caused my current issue. With a recent version of master including this change (commit 66d24c5) I am getting a regression in my code as demonstrated below:

PyTorch 0.2
```python
In [1]: import torch
In [2]: import numpy as np

In [3]: np.array([torch.LongTensor([1, 2]), torch.LongTensor([3, 4])])
Out[4]: 
array([
 1
 2
[torch.LongTensor of size 2]
,
       
 3
 4
[torch.LongTensor of size 2]
], dtype=object)

```
PyTorch master
```python
In [1]: np.array([torch.LongTensor([1, 2]), torch.LongTensor([3, 4])])
In [1]: import torch

In [2]: import numpy as np

In [3]: np.array([torch.LongTensor([1, 2]), torch.LongTensor([3, 4])])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-c15fb94d2d75> in <module>()
----> 1 np.array([torch.LongTensor([1, 2]), torch.LongTensor([3, 4])])

~/anaconda3/lib/python3.6/site-packages/torch/tensor.py in __int__(self)
    375         if self.numel() == 1:
    376             return int(self[(0,) * self.ndimension()])
--> 377         raise TypeError("only 1-element tensors can be converted "
    378                         "to Python scalars")
    379 

TypeError: only 1-element tensors can be converted to Python scalars
```

The reason to do this is so that I can say take three numpy arrays of torch tensors where the order lines up and:
1. Create a random ordering of batches with `batch_order = np.random.permutation(n_batches)`
2. During each epoch index all three numpy arrays like `x_array[batch_order]`, `other_input_array[batch_order]`, `y_array[batch_order]`. Each torch tensor represents one batch

Thanks!

**Pull request merged**